File: Classes/TopScoresSingleDetailViewController.m
//
//  TopScoresSingleDetailViewController.m
//  KidsTimeFun
//
//  Created by Jagmeet Chawla on 4/12/09.
//  Revised by Krishna Narayan on 9/17/09
//		--Updated graphics
//		--Added tell a friend
//		--Changed app name to Kids Learn To Tell Time
//  Copyright 2009-2013 NSC Partners LLC. Copyright 2014 One Step Ahead Apps, LLC. All rights reserved.
//

#import "TopScoresSingleDetailViewController.h"
#import "KidsTimeFunAppState.h"

@implementation TopScoresSingleDetailViewController

@synthesize activity;
@synthesize activityType;
@synthesize activityLevel;
@synthesize scoresArray;

- (id)initWithActivity:(int)act andType:(int)type andLevel:(int)level withNibName: (NSString *)nibNameOrNil bundle: (NSBundle *)nibBundleOrNil {
	self.activity = act;
	self.activityType = type;
	self.activityLevel = level;
	return [self initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
}

 // The designated initializer.  Override if you create the controller programmatically and want to perform customization that is not appropriate for viewDidLoad.
- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        // Custom initialization
    }
    return self;
}

// Implement viewDidLoad to do additional setup after loading the view, typically from a nib.
- (void)viewDidLoad {
	[self loadScoreForActivity:self.activity andType:self.activityType andLevel:self.activityLevel];
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning {
	// Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
	
	// Release any cached data, images, etc that aren't in use.
}

- (void)viewDidUnload {
	// Release any retained subviews of the main view.
	// e.g. self.myOutlet = nil;
}


#pragma mark Table view methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


// Customize the number of rows in the table view.
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return (([scoresArray count]>0)?[scoresArray count]:1);
}

// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Activity";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
//		cell = [[[UITableViewCell alloc] initWithFrame:CGRectZero reuseIdentifier:CellIdentifier] autorelease];
        cell = [[[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier] autorelease];
		if ([scoresArray count] > 0) {
			NSDictionary *score = [scoresArray objectAtIndex:indexPath.row];
			NSMutableString *cellText = [[NSMutableString alloc] init];
			[cellText appendFormat:@"%i",(int) (indexPath.row+1)];
			[cellText appendString:@". "];
			[cellText appendString:[score valueForKey:kPlayerName]];
			[cellText appendString:@", "];
			if (self.activityType == kActTypeNumbered) {
				[cellText appendFormat:@"%1.0f%%", [[score valueForKey:kPercentScore] floatValue]*100.0f];
				[cellText appendString:@", "];
				[cellText appendFormat:@"%i right",[[score valueForKey:kRightAnswers] intValue]];
				[cellText appendString:@", "];
				[cellText appendFormat:@"%i wrong", [[score valueForKey:kWrongAnswers] intValue]];
				[cellText appendString:@", "];
			}
			else {
				[cellText appendFormat:@"%i questions", [[score valueForKey:kRightAnswers] intValue] + [[score valueForKey:kWrongAnswers] intValue]];
				[cellText appendString:@", "];
			}
			[cellText appendFormat:@"%i sec",[[score valueForKey:kSecondsTaken] intValue]];
			cell.textLabel.text = [NSString stringWithString:cellText];
			[cellText release];
		}
		else cell.textLabel.text = kStrBlank ;
		if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) 
		{
			cell.textLabel.font = [UIFont fontWithName:@"Helvetica" size:24.0f];
		}
		else 
		{
			cell.textLabel.font = [UIFont fontWithName:@"Helvetica" size:15.0f];	
		}
	}
    cell.textLabel.textColor = [UIColor colorWithRed:0.043 green:0.376 blue:0.996 alpha:1.0];
    [cell setBackgroundColor:[UIColor colorWithWhite:1.00 alpha:0.25]];
	// Configure the cell.
    return cell;
}

- (void) loadScoreForActivity: (int)activity andType: (int)activityType andLevel: (int)activityLevel {
	//create scores Array
	//Now get the file name - root for documents directory + "/Scores/"
	NSString *docPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
	NSString *filePath = [docPath stringByAppendingString:@"/"];
	NSString *fileName = [NSString stringWithFormat:kFileVarScores,self.activity,self.activityType,self.activityLevel];
	NSString *fileNameWithPath = [filePath stringByAppendingString:fileName];
	//Now read scores array from disk
	scoresArray = [[NSArray alloc] initWithContentsOfFile:fileNameWithPath];
}

- (void)dealloc {
	[scoresArray release];
    [super dealloc];
}


@end

File: Classes/TopScoresSingleDetailViewController.swift
import UIKit

class TopScoresSingleDetailViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    var activity: Int32 = 0
    var activityType: Int32 = 0
    var activityLevel: Int32 = 0
    var scoresArray: [[String: Any]] = []

    convenience init(activity act: Int32, andType type: Int32, andLevel level: Int32,
                     withNibName nib: String?, bundle: Bundle?) {
        self.init(nibName: nib, bundle: bundle)
        activity = act; activityType = type; activityLevel = level
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        loadScores()
    }

    func loadScores() {
        let docs = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let fileName = String(format: kFileVarScores, activity, activityType, activityLevel)
        let path = docs + "/" + fileName
        scoresArray = (NSArray(contentsOfFile: path) as? [[String: Any]]) ?? []
    }

    func numberOfSections(in tableView: UITableView) -> Int { 1 }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return max(scoresArray.count, 1)
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let id = "Activity"
        let cell = tableView.dequeueReusableCell(withIdentifier: id) ??
            UITableViewCell(style: .default, reuseIdentifier: id)

        if !scoresArray.isEmpty {
            let score = scoresArray[indexPath.row]
            var text = "\(indexPath.row + 1). "
            text += (score[kPlayerName] as? String) ?? ""
            text += ", "
            if activityType == kActTypeNumbered {
                let pct = (score[kPercentScore] as? NSNumber)?.floatValue ?? 0
                let right = (score[kRightAnswers] as? NSNumber)?.intValue ?? 0
                let wrong = (score[kWrongAnswers] as? NSNumber)?.intValue ?? 0
                text += String(format: "%1.0f%%, %i right, %i wrong, ", pct * 100, right, wrong)
            } else {
                let r = (score[kRightAnswers] as? NSNumber)?.intValue ?? 0
                let w = (score[kWrongAnswers] as? NSNumber)?.intValue ?? 0
                text += "\(r + w) questions, "
            }
            text += "\((score[kSecondsTaken] as? NSNumber)?.intValue ?? 0) sec"
            cell.textLabel?.text = text
        } else {
            cell.textLabel?.text = kStrBlank
        }

        let fontSize: CGFloat = UIDevice.current.userInterfaceIdiom == .pad ? 24 : 15
        cell.textLabel?.font = UIFont(name: "Helvetica", size: fontSize)
        cell.textLabel?.textColor = UIColor(red: 0.043, green: 0.376, blue: 0.996, alpha: 1)
        cell.backgroundColor = UIColor(white: 1, alpha: 0.25)
        return cell
    }
}
